### DISCUSSION TITLE
epsilon_input_file with .hdf5 file

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#2376] [Q&A] [answered]
Hi everybody,

Trying to import custom 2D geometry using `epsilon_input_file` with an hdf5 file; hoping to get some guidance on the proper usage.

MWE:
```
%matplotlib inline
from matplotlib import pyplot as plt
import meep as mp
import h5py

sim = mp.Simulation(
    cell_size=mp.Vector3(1,1),
    resolution=64,
    epsilon_input_file='A.hdf5',
)

plt.figure(dpi=100)
sim.plot2D()
plt.show()
```

This produces a blank plot (no geometry). I've uploaded the source hdf5 file [here](https://github.com/dalarev/MEEP_scripts/blob/main/A.hdf5). It is simply a cylinder of radius `0.1a` with `eps = 2.5` and an air background.

I've tried both `.hdf5` and `.h5` files, but they produce the same results. I generated this array in Julia and exported with HDF5.jl.

I've confirmed that the file contains the expected geometry/dielectric values by visualizing with jupyterlab-h5web. 

I am just becoming familiar with hdf5 files, but I suspect the issue may be in the "structure" of the file: when I visualize using jupyterlab-h5web, I see a root and a child; perhaps this is in conflict with MEEP's guidance of being the "first/only" dataset in the file.

I'll also mention that if I give the `epsilon_input_file` some bogus name like `does_not_exist.hdf5`, no error is reported and the result is the same blank plot.

Looking forward to your inputs. Thank you!

**Comment by dalarev** [comment]
The issue persists even when I generate an .h5 file from MEEP geometry. 

MWE is below and notebook with outputs is [here](https://github.com/dalarev/MEEP_scripts/blob/main/test_hdf5.ipynb).

```
%matplotlib inline
from matplotlib import pyplot as plt
import numpy as np
import meep as mp

## Test: create MEEP geometry, output_epsilon, and try to read this MEEP-generated hdf5

resolution = 64
cell_size = mp.Vector3(1,1,0)

geometry = [mp.Cylinder(0.2, material=mp.Medium(epsilon=2.0))]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    geometry=geometry,
                   )
sim.plot2D()
sim.run(mp.at_beginning(mp.output_epsilon), until=0)

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    epsilon_input_file="eps-000000000.h5",
                   )
sim.plot2D()
plt.show()
```

**Comment by oskooi** [comment]
I think the issue is with the formatting of your HDF5 file. As a test, I used [h5utils](https://github.com/NanoComp/h5utils) and was not able to read it:
```
$ h5ls A.hdf5
A.hdf5: unable to open file
```
Can you try creating your HDF5 file using [h5py](https://docs.h5py.org/en/stable/quick.html)?

**Reply by dalarev** [reply]
Thanks @oskooi. 

The HDF5 file works fine for me using h5py. See [notebook output](https://github.com/dalarev/MEEP_scripts/blob/main/test_h5_file.ipynb). 

Regardless, in my [2nd post](https://github.com/NanoComp/meep/discussions/2376#discussioncomment-4718702), I generate the h5 file from within MEEP, which should be self-consistent but still fails.

Thanks for the help.

**Reply by dalarev** [reply]
Creating the HDF5 file using h5py is also producing a blank plot: notebook with output [here](https://github.com/dalarev/MEEP_scripts/blob/main/create_h5file_using_h5py.ipynb).

At this point, the only remaining user error I can think of is an incorrect value `epsilon_input_file`.

Appreciate your support with this.

**Comment by oskooi** [comment, marked as answer]
Looks like a bug in #1827 whereby loading the geometry via the `epsilon_input_file` feature still requires you to manually initialize the `meep.Simulation` object *before* invoking `plot2D`.

```
sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    epsilon_input_file="eps-000000000.h5",
                   )

# manually initialize the sim object
sim.init_sim() 

sim.plot2D()
plt.show()
```

With this change, the simulation output does show that the HDF5 file was successfully read.
```
Initializing structure...
read in 64x64x1 epsilon-input-file "eps-000000000.h5"
time for choose_chunkdivision = 8.069e-06 s
read in 64x64x1 epsilon-input-file "eps-000000000.h5"
Working in 2D dimensions.
Computational cell is 1 x 1 x 0 with resolution 64
time for set_epsilon = 0.00823663 s
-----------
read in 64x64x1 epsilon-input-file "eps-000000000.h5"
```
![plot2D_epsilon_input_file](https://user-images.githubusercontent.com/7152530/214154163-47b094ba-23fc-47cf-834a-a28ccbd37e29.png)

Thanks for catching this. Will try to post a fix soon. In the meantime, the workaround is quite straightforward.

**Reply by dalarev** [reply]
Awesome, thanks for the notes!

I also just learned that defining `default_material` as a NumPy array works, too.

Thanks again!
